model/lollipop_keys.ts

```diff
@@ -0,0 +1,190 @@
+]);
+export type RetrievedLolliPopPubKeys = t.TypeOf<
+  typeof RetrievedLolliPopPubKeys
+>;
+
+export class LolliPOPKeysModel extends CosmosdbModelVersionedTTL<
+  LolliPopPubKeys,
+  NewLolliPopPubKeys,
+  RetrievedLolliPopPubKeys,
+  typeof LOLLIPOPKEYS_MODEL_ID_FIELD
+> {
+  constructor(container: Container) {
+    super(
+      container,
+      NewLolliPopPubKeys,
+      RetrievedLolliPopPubKeys,
+      LOLLIPOPKEYS_MODEL_ID_FIELD
+    );
+  }
+
+  /*
+   * Reserve the key by creating a new document with version 0 with the ttl setted for the time needed,
+   * */
+  public create(
+    popDocument: LolliPopPubKeys,
+    option?: RequestOptions
+  ): TE.TaskEither<CosmosErrors, RetrievedLolliPopPubKeys> {
+    return pipe(
+      this.getTtlValue(popDocument),
+      // super.create never returns 409 error but a generic CosmosErrorResponse with io-functions-commons v26.8.1
+      TE.chain(ttl => super.create({ ...popDocument, ttl }, option))
+    );
+  }
+
+  /*
+   * Update the last version of the document setting the new properties and the ttl at 2 years
+   * */
+  public upsert(
+    popDocument: LolliPopPubKeys,
+    option?: RequestOptions
+  ): TE.TaskEither<CosmosErrors, RetrievedLolliPopPubKeys> {
+    return pipe(
+      this.getTtlValue(popDocument),
+      TE.chain(ttl => super.upsert({ ...popDocument, ttl }, option))
+    );
+  }
+
+  /**
+   * This method is disabled to avoid wrong use cases. Use upsert instead.
+   *
+   * @deprecated
+   * */
+  public update(
+    _: RetrievedLolliPopPubKeys
+  ): TE.TaskEither<CosmosErrors, never> {
+    return TE.left(
+      toCosmosErrorResponse(new Error("Cannot update a lollipop document"))
+    );
+  }
+
+  /**
+   * This method is disabled to avoid wrong use cases.
+   *
+   * @deprecated Use updateKeys instead.
+   * */
+  public updateTTLForAllVersions(): TE.TaskEither<CosmosErrors, never> {
+    return TE.left(
+      toCosmosErrorResponse(
+        new Error("Cannot update tll for old versions in a lollipop document")
+      )
+    );
+  }
+
+  private getTtlValue(
+    popDocument: LolliPopPubKeys
+  ): TE.TaskEither<CosmosErrors, NonNegativeInteger> {
+    return pipe(
+      super.findLastVersionByModelId([popDocument.assertionRef]),
+      TE.map(
+        flow(
+          // if the last version was PENDING the new ttl is setted to TTL_VALUE_AFTER_UPDATE
+          // if the last version ttl is missing then the new ttl is setted to TTL_VALUE_AFTER_UPDATE to avoid setting the ttl to a negative value
+          O.map(lastPop =>
+            lastPop.status === PubKeyStatusEnum.PENDING ||
+            (lastPop.ttl ?? 0) < 1
+              ? TTL_VALUE_AFTER_UPDATE
+              : // eslint-disable-next-line @typescript-eslint/restrict-plus-operands, no-underscore-dangle
+                ((lastPop._ts +
+                  (lastPop.ttl ?? 0) -
+                  Math.floor(
+                    new Date().getTime() / 1000
+                  )) as NonNegativeInteger)
```

Comment:
What if an expired document reach this code? This will not be a NonNegativeInteger. Should we worry about this, or is it just impossible?

Reply:
IHMO this is not a possible scenario, an expired document means that it should be deleted by the `ttl` logic

Reply:
This is not possible because after the create or upsert operation a decode operation is executed, and an eventual negative ttl never be saved on cosmos. A `CosmosDecondingError` is provided.